File: Selenium/Activities/Activity21.py
import time

from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support import expected_conditions
from selenium.webdriver.support.select import Select
from selenium.webdriver.support.wait import WebDriverWait
from webdriver_manager.firefox import GeckoDriverManager
from selenium.webdriver.firefox.service import Service as FirefoxService

serv = FirefoxService(GeckoDriverManager().install())

with webdriver.Firefox(service=serv) as driver:
    driver.get("https://www.training-support.net/selenium/tab-opener")
    main_window = driver.current_window_handle
    print("Current handle", main_window)
    wait = WebDriverWait(driver, 10)
    driver.find_element(By.ID, "launcher").click()
    wait.until(expected_conditions.number_of_windows_to_be(2))
    print("All window handles: ", driver.window_handles)
    driver.switch_to.window(driver.window_handles[1])
    print("Title of new tab: ", driver.title)
    time.sleep(5)
    heading = driver.find_element(By.XPATH, "//div[@class='content']")
    wait.until(expected_conditions.visibility_of(heading))
    print("Content:", heading.text)
    driver.find_element(By.ID, "actionButton").click()
    wait.until(expected_conditions.number_of_windows_to_be(3))
    driver.switch_to.window(driver.window_handles[2])
    time.sleep(5)
    print("Title of new tab: ", driver.title)
    print("Content:", driver.find_element(By.CLASS_NAME, "content").text)
    driver.quit()

File: Selenium/Activities/Activity21.java
package Activities;
import io.github.bonigarcia.wdm.WebDriverManager;
import org.openqa.selenium.Alert;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.support.ui.ExpectedCondition;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.Select;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.time.Duration;
import java.util.List;

public class Activity21 {
    public static void main(String[] args) throws InterruptedException {
        WebDriverManager.firefoxdriver().setup();
        System.setProperty(FirefoxDriver.SystemProperty.BROWSER_LOGFILE, "/dev/null");
        WebDriver newDriver = new FirefoxDriver();
        WebDriverWait wait = new WebDriverWait(newDriver, Duration.ofSeconds(10));
        newDriver.get("https://www.training-support.net/selenium/tab-opener");
        System.out.println("Title of page: " + newDriver.getTitle());
        System.out.println("Current window: "+newDriver.getWindowHandle());
        newDriver.findElement(By.id("launcher")).click();
        wait.until(ExpectedConditions.numberOfWindowsToBe(2));
        System.out.println("Current open windows: " + newDriver.getWindowHandles());
        for(String handle : newDriver.getWindowHandles()) {
            newDriver.switchTo().window(handle);
        }
        wait.until(ExpectedConditions.elementToBeClickable(By.id("actionButton")));
        System.out.println("Page title: " + newDriver.getTitle());
        String pageHeading = newDriver.findElement(By.className("content")).getText();
        System.out.println("Page Heading: " + pageHeading);
        newDriver.findElement(By.id("actionButton")).click();

        wait.until(ExpectedConditions.numberOfWindowsToBe(3));
        for(String handle : newDriver.getWindowHandles()) {
            newDriver.switchTo().window(handle);
        }

        wait.until(ExpectedConditions.visibilityOfElementLocated(By.className("content")));
        System.out.println("Current tab: " + newDriver.getWindowHandle());
        System.out.println("Page title: " + newDriver.getTitle());
        pageHeading = newDriver.findElement(By.className("content")).getText();
        System.out.println("Page Heading: " + pageHeading);
        newDriver.quit();

    }
}